### PR TITLE
Obtain patch level metadata (e.g. geospatial bounds and cloud cover), save and demo DEP use case (sim search)

### PR DESCRIPTION
**Pull request by lillythomas**
This PR will work towards a demonstration of how to obtain patch level embeddings and write them to GeoParquet files to run similarity search with. 

Main tasks that need to be done:

- [x] Given the geospatial bounds of a 512x512 parent chip, obtain the child 32x32 geospatial bounding boxes (patches)
- [x] Map the child bounding boxes to patch level embeddings (making sure `embeddings_level = "patch"`)
- [x] Create geodataframes for the patch embeddings with their mapped bbox and write to GeoParquet
- [x] Load in a lanceDB
- [x] Use the ground truth points from #140 https://github.com/Clay-foundation/model/discussions/140#discussioncomment-8437613  [mineral extraction training data](https://github.com/digitalearthpacific/mineral-resource-detection/blob/main/training_data/draft_inputs/MRD_dissagregated_25.geojson) as reference for similarity search against all patch embeddings
- [x] compute and save patch level cloud cover percentages (+ add to db table for sim search)

Reference tickets: #168 #140

### CODE REVIEW

**Comment by lillythomas** [comment]
The notebook `docs/tutorial_digital_earth_pacific_patch_level.ipynb` walks through an example of:
- generating patch level embeddings for an area where known mining extraction events occur
- saving the patch level embeddings to independent GeoParquet files
- executing similarity search based on a ground truth point's overlapping patch

@weiji14 @yellowcap ready for when you have time to review.

**Comment by weiji14** [comment]
Thanks @lillythomas! I haven't looked too closely yet, but would it be possible to show where the similarity search results are located? Maybe something like showing the bounding boxes of all the patches on a map, and also overlay where the original quarry points are.

**Comment by lillythomas** [comment]
> Thanks @lillythomas! I haven't looked too closely yet, but would it be possible to show where the similarity search results are located? Maybe something like showing the bounding boxes of all the patches on a map, and also overlay where the original quarry points are.

Yes! Great idea. Working on this tomorrow.

**Review by yellowcap — COMMENTED** [review]
Thanks Lilly, this is great to see! I did not yet have time get to the sim search part, but will get back to it on Monday. Left some comments for now.

Could you add some more context on what the purpose is? Do you think we going to apply this to a region for searching certain events / features?

File: docs/tutorial_digital_earth_pacific_patch_level.ipynb
Comment (lillythomas) [reply docs/tutorial_digital_earth_pacific_patch_level.ipynb:857]: This was introduced by making your suggested change in https://github.com/Clay-foundation/model/pull/172#discussion_r1526415973 but I have a way to handling it. Essentially pyarrow doesn't like a polygon object, so I can get a list equivalent for the table by adding the bounds method to `"box": row["box"]` e.g. `"box": row["box"].bounds`, which I am in favor of doing as it allows us to drop the lamba function as you pointed out.

File: src/model_clay.py
Comment (lillythomas) [reply src/model_clay.py:793]: True. Will do.

**Comment by lillythomas** [comment]
@weiji14 @yellowcap when you have a moment, this could use another review before merging.

File: docs/tutorial_digital_earth_pacific_patch_level.ipynb
Comment (weiji14) [inline docs/tutorial_digital_earth_pacific_patch_level.ipynb:369]: Since we're working with optical imgaery, it is recommend to set pixel registation to center registered, rather than `stackstac`'s default topleft (gridline) registration. See https://stackstac.readthedocs.io/en/latest/api/main/stackstac.stack.html#stackstac.stack.params.xy_coords.

```suggestion
    "    resampling=Resampling.nearest,\n",
    "    xy_coords=\"center\",\n",
```

**Review by weiji14 — COMMENTED** [review]
Could you please add this tutorial notebook to `docs/_toc.yml`, so that the notebook will render on merge? Also made some suggested changes/typos below.

One major issue I saw is on conflating Digital Earth Pacific's cloud-free GeoMAD composite with a clouded Sentinel-2 image. Matching SCL metadata from raw Sentinel-2 images to the GeoMAD composite doesn't seem correct...

File: docs/tutorial_digital_earth_pacific_patch_level.ipynb
Comment (weiji14) [inline docs/tutorial_digital_earth_pacific_patch_level.ipynb:18]: Mixing the cloud-free composite from DEP'S STAC with the SCL mask from AWS's Sentinel 2 STAC seems... weird? I see that you're trying to consolidate the notebook from #184 with the one here, but there's an inherent mismatch with the data from two different STAC catalogs. It won't make sense to do a cloud-percentage pre-filter using the SCL band, when the composite image is technically cloud-free already.

File: docs/tutorial_digital_earth_pacific_patch_level.ipynb
Comment (weiji14) [inline docs/tutorial_digital_earth_pacific_patch_level.ipynb:19]: Mixing the cloud-free composite from DEP'S STAC with the SCL mask from AWS's Sentinel 2 STAC seems... weird? I see that you're trying to consolidate the notebook from #184 with the one here, but there's an inherent mismatch with the data from two different STAC catalogs. It won't make sense to do a cloud-percentage pre-filter using the SCL band, when the composite image is technically cloud-free already.

Comment (yellowcap) [reply docs/tutorial_digital_earth_pacific_patch_level.ipynb:19]: For this example notebook, could we use a regular Sentinel 2 scene from AWS? We could hand pick one that does not have many clouds, but maybe a few as to demonstrate the cloud percentage calculation?

In other words, what is the advantage of using the DEP STAC in this example?

Another solution could be to use a light weight cloud detection model that could compute a cloud mask based on the bands directly. There are a few, but not sure how easy they are to use.

**Comment by lillythomas** [comment]
Yes @weiji14 @yellowcap I didn't intend to suggest that we would use the derived cloud percentages to filter the embedding results obtained from cloud-free DEP composite, just that the two leveraged some similar things (i.e. AOI) and the initial ask for patch-level metadata https://github.com/Clay-foundation/model/issues/168 was unified so I sought to consolidate.

I guess indeed if we want this to be a sensible consolidation, we could demonstrate this without the cloud-free composite or I can just re-separate the PRs. Whatever is swiftest and most useful. I'll address some of the other comments from Wei Ji first then comment on which of the two directions we go with as far as the above is concerned.

**Comment by lillythomas** [comment]
Re-separated into different tutorials / PRs. See https://github.com/Clay-foundation/model/pull/184 for the cloud cover tutorial.

**Comment by yellowcap** [comment]
I was able to run the notebook, works nicely now! Only one celll copmlained, the one with the plot of the similar patches, I get the following error.

```
ValueError                                Traceback (most recent call last)
Cell In[53], line 41
     39 idx = row["idx"]
     40 # Find the corresponding window based on the idx
---> 41 window_index = idxs_windows["idx"].index(idx)
     42 window_data = idxs_windows["window"][window_index]
     43 # print(window_data.shape)

ValueError: '20180828_4_2' is not in list
```

I really like the last plot showing the location of the search results. I would have one request there: would it be possible to highlight which one was the input patch? For the user to know what the reference was.

I thought of this because in the plot none of the patches have a ground truth point right in the center. At least the reference chips should be right on top of a quarry ideally.

**Comment by lillythomas** [comment]
Thanks, glad you could reproduce @yellowcap. As for that error you're encountering, we shouldn't have any dates in 2018 in this tutorial (see `daterange: dict = ["2021-01-01T00:00:00Z", "2021-12-31T23:59:59Z"] `)  so maybe just double check `data/minicubes` is cleared before running the notebook. Other than that, I'll adjust the plot to support that request; I agree it would be useful.

**Comment by lillythomas** [comment]
@yellowcap specifying the reference patch in the final plot was achieved in commit  https://github.com/Clay-foundation/model/commit/1f9bdf53f3a3811f08aeb7f205f2d1cfd2df0dc1.

> At least the reference chips should be right on top of a quarry ideally.

The reference patch overlaps directly with a ground truth point. It is the box in yellow here:
Screen Shot 2024-03-21 at 5 17 41 PM

**Comment by yellowcap** [comment]
Looks like the linter is not happy about the size of the notebook.

I would propose to replace the cells with images with links that will make the notebook smaller.

I did this by uploading the images from the cells to the PR or Issue, then copy the link and place it in a markdown cell. The following is an example:
```
![Minicube visualization](https://github.com/Clay-foundation/model/assets/901647/c6e924e5-6ba1-4924-b99a-df8b90731a5f)
```

See https://github.com/Clay-foundation/model/blob/main/docs/partial-inputs.ipynb?short_path=984f94c#L181

**Review by yellowcap — APPROVED** [review]
This is good to go once the linters are happy! 🚀

**Pull request merged**
